### ISSUE TITLE
Overly verbose logging when migrations directory specified

### ISSUE THREAD

**Issue by anandnimkar** [knex/knex#3921]
# Environment

Knex version: 0.21.2
Database + version: Irrelevant
OS: Irrelevant

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do

Due to changes introduced in https://github.com/knex/knex/pull/3839, on every call to `migrate.latest()` I now get two messages in my logs stating `"FS-related option specified for migration configuration. This resets migrationSource to default FsMigrations"`

I think this is overly verbose and possibly unnecessary. I run migrations programmatically at the start of every invocation of a Lambda function and now this is showing up in my logs.

At the very least, can we just have this message once per migrate.latest() call? Because `getMergedConfig` (in the file `lib/migrate/configuration-merger.js`) is called multiple times, it is showing up twice.

I also introduced a separate issue to ensure that `getMergedConfig` is called with the appropriate client logger when supplied. See https://github.com/knex/knex/issues/3919 and https://github.com/knex/knex/pull/3920.

2. Error message

See 1 above.

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to MySQL or PostgreSQL, then single file example which initializes
   needed data and demonstrates the problem.

Set migrations directory in knexfile, then run `migrate.latest()` or `migrate.rollback()`

**Comment by ksmithut** [comment]
I might also add that I'm fine if there's a more preferred method of setting custom migration directories (perhaps setting migrationSource to an instance of `FsMigrations`), but it would be nice if it were documented (and supported in the types).

**Comment by elhigu** [comment]
You can set migration directory in configuration (https://knexjs.org/#Migrations-API)

```
{
  migrations: {
    directory: ['myCustomMigrationDir1','myCustomMigrationDir2']
  }
}
```

**Comment by ksmithut** [comment]
@elhigu Right, the problem is that when using the `knex` js api for running migrations, you get verbose logging `"FS-related option specified for migration configuration. This resets migrationSource to default FsMigrations"` when using that option using the latest version

**Comment by zjr** [comment]
Same if running `knex migrate` from the command line. I get the message twice.

**Comment by jonahmolinski** [comment]
@jalfd exactly!

There should be a way to suppress this message or clear documentation on how to avoid it.

**Comment by elhigu** [comment]
> @jalfd exactly!
> 
> There should be a way to suppress this message or clear documentation on how to avoid it.

Overriding loggers can be used to filter error messages. So additional configuration switches for supressing specific warnings has not been added for a long time.

**Comment by jonahmolinski** [comment]
> Overriding loggers can be used to filter error messages (really useful in tests). So additional configuration switches for supressing specific warnings has not been added for a long time. That might not be clear from docs though.

Maybe I am missing something, but I pass in a logger via the "log" parameter in config and it's not used for this message. It is used for the debug SQL statements however. So it must be working...

How can one set the logger that is used for that message? Cheers!

**Comment by elhigu** [comment]
> > Overriding loggers can be used to filter error messages (really useful in tests). So additional configuration switches for supressing specific warnings has not been added for a long time. That might not be clear from docs though.
> 
> Maybe I am missing something, but I pass in a logger via the "log" parameter in config and it's not used for this message. It is used for the debug SQL statements however. So it must be working...
> 
> How can one set the logger that is used for that message? Cheers!

You need to use pattern matching in logger function to check log messages  to filter out. Docs how to set loggers is here https://knexjs.org/#Installation-log

**Comment by joshkel** [comment]
It looks like it's straightforward and reasonable to modify Knex to only log if the migrationSource is actually reset; see #3936.

**Comment by StephanHoyer** [comment]
Monkey patch for the logger as a work around.

```js
const db = knex(dbConfig)

const VERBOSE_WARNING_ABOUT_MIGRATION_CONFIGURATION =
  'FS-related option specified for migration configuration. This resets migrationSource to default FsMigrations'

const _warn = db.context.client.logger.warn
db.context.client.logger.warn = m => {
  if (m !== VERBOSE_WARNING_ABOUT_MIGRATION_CONFIGURATION) {
    return _warn.call(db.context.client.logger, m)
  }
}
```

**Comment by elhigu** [comment]
@StephanHoyer No need to monkeypatch. Knex configuration has a support for overriding logger as I already mentioned on my last comment.

**Comment by kevholder** [comment]
@elhigu can you share the code to fix this via logger configuration so everyone that googles this has a simple solution?

**Comment by kibertoad** [comment]
I'll release a proper fix for this today.

**Comment by kibertoad** [comment]
Fix released in 0.21.6, sorry for the delay.
